### PR TITLE
Go 1.11 module support

### PR DESCRIPTION
**Pull request by talbright**
**MODULE SUPPORT**

> Go 1.11 includes preliminary support for versioned modules as proposed here. Modules are an experimental opt-in feature in Go 1.11, with the hope of incorporating feedback and finalizing the feature for Go 1.12. Even though the details may change, future releases will support modules defined using Go 1.11 or vgo.

This PR drops the `v3` path. For detailed discussion see comments in this PR. 

This PR continues support for vendoring for older versions of GO (https://github.com/golang/go/wiki/Modules#how-do-i-use-vendoring-with-modules-is-vendoring-going-away)

> Older versions of Go such as 1.10 understand how to consume a vendor directory created by go mod vendor, so vendoring is one way to provide dependencies to older versions of Go that do not fully understand modules.

**TODO**

* [x] Determine vendoring approach
* [x] Update docs 
* [x] Improve build with Makefile(s) which: will be invoked by travis instead of `go test` and include appropriate checks such as gofmt
* [x] Update travis to build 1.9 - 1.12 (https://docs.travis-ci.com/user/customizing-the-build/#build-matrix) 
* [x] Minor improvements to schema generation script and inclusion of generation target in makefile

⭐️  After merging be sure to create a github release
⭐️  Use semver and github releases moving forward 

Addresses #37

### CODE REVIEW

**Comment by cyberdelia** [comment]
For the record, `v3` was meant to match the API version more than indicating the version of the “library” itself.

**Comment by alindeman** [comment]
> For the record, `v3` was meant to match the API version more than indicating the version of the “library” itself.

I think this is important. Is there any way we can make this project compatible with modules without bumping to v4?

**Comment by talbright** [comment]
Unfortunately `v1`, `v2`, ... in GO projects has a special meaning that doesn't seem to be related to the use case here (https://github.com/golang/go/wiki/Modules#semantic-import-versioning) and is causing some conflicts with the new module system. 

I can work on a different approach but that may reduce compatibility. Is that a preferred tradeoff?

**Comment by mars** [comment]
> Unfortunately v1, v2, ... in GO projects has a special meaning that doesn't seem to be related to the use case here

😬 wow that's a collision of intents for sure.

I always assumed that `v3` indicated the Heroku API `v3`, but if it's conflated with Go import versioning, perhaps we should rename that troubled module path to `version3` or `api-v3` as part of this major change.

**Comment by talbright** [comment]
> 😬 wow that's a collision of intents for sure.
> 
> I always assumed that `v3` indicated the Heroku API `v3`, but if it's conflated with Go import versioning, perhaps we should rename that troubled module path to `version3` or `api-v3` as part of this major change.

The main concern I have with doing that (or my preference -- flattening the directory structure), is breaking existing users. I would be less concerned if there had been releases/tagging. 

What's the purpose of this pattern? Is it really needed? This client does not utilize variants, so it's always the official supported API "latest GA". Does Heroku (or has it ever) supported multiple GA versions of its platform API?

**Comment by mars** [comment]
Before merging any module system changes, let's definitely please 🙏 make a release containing the current "pre-go-modules" client.

I'm happy to go ahead with that, call it `v1.0.0` to kick-off our semantic versioning, and clearly describe the release as the "Initial stable release before switching to the new Go 1.11 modules system".

**Comment by fgblomqvist** [comment]
Any update on this? Also, any recommended way of importing the package (as it currently is) into a go mod project?

Right now, the only way I can get working is to `require github.com/heroku/heroku-go v0.0.0-20190103224148-ad17585a922f` and then import `github.com/heroku/heroku-go/v3` inside my code. Doesn't seem like the proper way. I tried doing a `replace github.com/heroku/heroku-go => github.com/heroku/heroku-go/v3 latest` which seemed to initially work (it would tag it as v3.0.0.....) but then it just complained about not finding a go.mod in the root or in the v3. But yeah, this is the first package I've dealt with that uses the directory structure-way of versioning it.

**Comment by talbright** [comment]
@fgblomqvist I'm waiting to here back from the team... 

@alindeman @cyberdelia any thoughts? Happy to make changes to this PR if there's a preferred approach.

**Comment by mars** [comment]
I created our first release 🙀 [v1.0.0](https://github.com/heroku/heroku-go/releases/tag/v1.0.0).

We'll move forward with this likely breaking PR soon.

**Comment by mars** [comment]
Hey @talbright ,

Let's drop `/v3` from the module path so usage becomes:

```
import (
  heroku "github.com/heroku/heroku-go"
)
```

It will be a breaking change of course. We'll tag it as `v2.0.0`, and add an obvious *Breaking Change* notice to the README.

Does anything else need to happen to resolve the problem using this as a Go Module?

Let me know if you have time to work on this, or if it would be helpful for me to take it from here 😇

**Comment by mars** [comment]
Apparently I have not read enough Go Module documentation. (Thanks to friendly collaborators pointing this out.)

According to ["Go Modules: Semantic Import Versioning" docs](https://github.com/golang/go/wiki/Modules#semantic-import-versioning) we don't have an option:

> If the module is version v2 or higher, the major version of the module must be included as a /vN at the end of the module paths used in go.mod files (e.g., module github.com/my/mod/v2, require github.com/my/mod/v2 v2.0.0) and in the package import path (e.g., import "github.com/my/mod/v2/mypkg").

The current version will actually need to be tagged `v3.0.0`, and the new Go Modules version will need its path to be changed to `/v4` and tagged `v4.0.0`.

**Comment by talbright** [comment]
> Hey @talbright ,
> 
> Let's drop `/v3` from the module path so usage becomes:
> 
> It will be a breaking change of course. We'll tag it as `v2.0.0`, and add an obvious _Breaking Change_ notice to the README.
> 
> (See revision to plan in next comment.)
> 
> Does anything else need to happen to resolve the problem using this as a Go Module?
> 
> Let me know if you have time to work on this, or if it would be helpful for me to take it from here 😇
I can follow through and finish up. There's some other nice changes for the build on the branch I want to make sure don't get lost as well.

I will review that doc again (its been a few months.) From what I recall though, if we aren't worried about breaking backwards compatibility we do have the option of dropping `vN`. I somewhat prefer this approach:

- it removes the `vN` vestiges and naming convention for which I'm not a fan of
- simplifies module support going forward
- makes the code base easier to maintain 

In the future if Heroku truly needs to maintain two separate versions of the API (the original reason the `vN` notation was used), likely the best thing to do is create a new project for it.

**Comment by mars** [comment]
Sounds great @talbright

**Comment by talbright** [comment]
@mars ready for review

**Comment by mars** [comment]
Nice @talbright! 🙌😄 I'll review and test how it works with the Terraform Provider next week.

File: .travis.yml [hunk 1]
```diff
@@ -1 +1,7 @@
 language: go
+
+go:
+  - "1.9"
```
Comment (mars) [inline .travis.yml]: Are Go Modules backward-compatible? These CI versions might need some pruning.

Comment (talbright) [reply .travis.yml]: Not exactly no. But vendoring is the magic that makes this work. Please see https://github.com/golang/go/wiki/Modules#how-do-i-use-vendoring-with-modules-is-vendoring-going-away for more details. Also note the matrix is building all these versions just fine so things seem to line up fine.

**Comment by talbright** [comment]
> Thanks for your patience as I finally got to this. Just a few comments in-line.
No worries, this PR is just taking its time to germinate. We will get there.

> 
> Also, I opened a [PR on the Terraform Provider](https://github.com/terraform-providers/terraform-provider-heroku/pull/195) that uses this branch:
> 
>     * validated that this Provider branch works correctly with Terraform, applying & destroying a simple configuration
> 
>     * currently running the full Provider test suite locally (will report back, but it's all success so far)
Awesome, thanks for doing this.

**Comment by mars** [comment]
✅ Hashicorp CI passed for [the Terraform Provider on this branch](https://github.com/terraform-providers/terraform-provider-heroku/pull/195).

**Comment by talbright** [comment]
w00t! BTW I don't have perms to merge...

**Comment by mars** [comment]
I'm releasing `v4.0.0` now!

**Pull request merged**